### ISSUE TITLE
Consider adding prometheus metrics for DMCI /insert, /update and maybe /validate endpoints

### ISSUE THREAD

**Issue by magnarem** [metno/discovery-metadata-catalog-ingestor#220]
The https://github.com/rycus86/prometheus_flask_exporter, supports exporting flask metrics out of the box.
This could be beneficial to monitor HTTP error codes for the DMCI endpoints, if insert and update calls fails etc, and then we can get alerts if DMCI have a lot of failed inserts for example.

**Comment by magnarem** [comment]
Also what my thought was for this issue, when we have a working prometheus exporter, is to see if we also can create some 
custom metrics for the worker in dmci.

The worker returns a list of failed distributors, so we should have some metrics like:

# failed file_dist <counter>
# failed csw_dist <counter>
# failed solr_dist <counter>

From the failed array in https://github.com/metno/discovery-metadata-catalog-ingestor/blob/main/dmci/api/worker.py#L171.

Maybe the code have to be rewritten a bit, since we dont get this array back into the /update and /insert endpoints,

So the lines at https://github.com/metno/discovery-metadata-catalog-ingestor/blob/main/dmci/api/app.py#L74
and https://github.com/metno/discovery-metadata-catalog-ingestor/blob/main/dmci/api/app.py#L69

need to get the failed array back. so the code

`msg, code = self._insert_update_method_post("insert", request)` need to be changed to
`msg, code, failed = self._insert_update_method_post("insert", request)`
and if there are elements in the failed array, then it should be possible to add it as a custom metric using the prometheus_flask_exporter API.

**Comment by charlienegri** [comment]
do I read correctly that the only place in practice where this 'failed' array would be not null is in this if block? https://github.com/metno/discovery-metadata-catalog-ingestor/blob/main/dmci/api/app.py#L159  
because it would have to be returned by `_distributor_wrapper` at L157, something like   
`err, failed = self._distributor_wrapper(worker)`  
and then it's relevant only for that next if block

**Comment by magnarem** [comment]
Yes. You are right. you will have to return the failed array from the `_distributor_wrapper` and then if `err` and `failed` is not empty then add to failed <dist> metrics.

I think if `failed` is not empty, the elements there are the name of the distributors that failed.

**Comment by charlienegri** [comment]
ok this probably means creating custom metrics counters defined via prometheus_client and incrementing them in `post_insert` and `post_update` based on failed, but I have no idea of that is out of the box compatible with the GunicornPromethusMetrics wrapper or then there's gonna have to have some extra custom collection registry layer...

**Comment by magnarem** [comment]
I am not sure myself how this is working.. but It should be possible somehow. I bet other projects using flask app also want to expose custom prometheus metrics for their business-logic and not just reponse codes from api endpoints

**Comment by charlienegri** [comment]
anyway, for now still no default metrics exposure..   
from the pod /proc/net/tcp I see nothing about port 9200    
port-forwarding also dos not help
I guess this approach is too naive.. or then I picked a cursed port

**Comment by charlienegri** [comment]
giving it a try https://gitlab.met.no/tjenester/s-enda/-/merge_requests/2928

**Comment by magnarem** [comment]
Seems like it works now:
https://dmci.s-enda-dev.k8s.met.no/metrics

But now I do not see those default `prometheus_flask_exporters` (https://github.com/rycus86/prometheus_flask_exporter?tab=readme-ov-file#default-metrics) for the endpoints anymore. Just the custom failed, and some python related metrics. Would be nice to have also, for checking request time and http codes

```
# HELP python_gc_objects_collected_total Objects collected during gc
# TYPE python_gc_objects_collected_total counter
python_gc_objects_collected_total{generation="0"} 976.0
python_gc_objects_collected_total{generation="1"} 569.0
python_gc_objects_collected_total{generation="2"} 0.0
# HELP python_gc_objects_uncollectable_total Uncollectable objects found during GC
# TYPE python_gc_objects_uncollectable_total counter
python_gc_objects_uncollectable_total{generation="0"} 0.0
python_gc_objects_uncollectable_total{generation="1"} 0.0
python_gc_objects_uncollectable_total{generation="2"} 0.0
# HELP python_gc_collections_total Number of times this generation was collected
# TYPE python_gc_collections_total counter
python_gc_collections_total{generation="0"} 154.0
python_gc_collections_total{generation="1"} 13.0
python_gc_collections_total{generation="2"} 1.0
# HELP python_info Python platform information
# TYPE python_info gauge
python_info{implementation="CPython",major="3",minor="10",patchlevel="12",version="3.10.12"} 1.0
# HELP process_virtual_memory_bytes Virtual memory size in bytes.
# TYPE process_virtual_memory_bytes gauge
process_virtual_memory_bytes 1.550450688e+09
# HELP process_resident_memory_bytes Resident memory size in bytes.
# TYPE process_resident_memory_bytes gauge
process_resident_memory_bytes 9.6800768e+07
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1.72837924519e+09
# HELP process_cpu_seconds_total Total user and system CPU time spent in seconds.
# TYPE process_cpu_seconds_total counter
process_cpu_seconds_total 3.06
# HELP process_open_fds Number of open file descriptors.
# TYPE process_open_fds gauge
process_open_fds 11.0
# HELP process_max_fds Maximum number of open file descriptors.
# TYPE process_max_fds gauge
process_max_fds 1.048576e+06
# HELP failed_file_dist_total Number of failed file_dist
# TYPE failed_file_dist_total counter
# HELP failed_csw_dist_total Number of failed csw_dist
# TYPE failed_csw_dist_total counter
# HELP failed_solr_dist_total Number of failed solr_dist
# TYPE failed_solr_dist_total counter
# HELP flask_exporter_info Information about the Prometheus Flask exporter
# TYPE flask_exporter_info gauge
flask_exporter_info{version="0.23.1"} 1.0
```

**Comment by charlienegri** [comment]
maybe we can add to the mid-level instance the dashboard of the exporter customized with out extra metrics, I 'll look into it

**Comment by charlienegri** [comment]
it's a start https://prometheus-fou-b.met.no/d/_eX4mpl3/prometheus-flask-exporter-custom-dashboard-for-dmci?orgId=1&from=now-15h&to=now&refresh=30s
